File: sync-client/packages/core/src/cache/metadata-cache.ts
import { join } from 'node:path';
import { existsSync, mkdirSync } from 'node:fs';
import { CACHE_DIRNAME, METADATA_CACHE_FILENAME } from './constants.js';
import { loadJsonIndex, saveJsonIndex } from './cache-utils.js';
import type { CachedPlaylistData, MetadataCacheData } from './types.js';
import type { FileInfo } from '../types.js';

/** Current schema version for the metadata cache file. */
const METADATA_CACHE_VERSION = 1;

/**
 * Persistent cache for API response metadata (playlist file lists + ETags).
 *
 * Stores one JSON file per profile at <configDir>/cache/<profile>/metadata-cache.json.
 * Separate from the audio cache index (cache-index.json).
 */
export class MetadataCache {
  private readonly cachePath: string;
  private data: MetadataCacheData;

  constructor(configDir: string, profile: string) {
    const cacheDir = join(configDir, CACHE_DIRNAME, profile);
    if (!existsSync(cacheDir)) {
      mkdirSync(cacheDir, { recursive: true });
    }
    this.cachePath = join(cacheDir, METADATA_CACHE_FILENAME);
    this.data = loadJsonIndex<MetadataCacheData>(
      this.cachePath,
      { profile, version: METADATA_CACHE_VERSION, playlists: {} },
      (d) => d.profile === profile && d.version === METADATA_CACHE_VERSION,
    );
  }

  // ── Read ──

  /** Get cached playlist data, or null if not cached. */
  getPlaylistFiles(playlistKey: string): CachedPlaylistData | null {
    return this.data.playlists[playlistKey] ?? null;
  }

  /** Get list of playlist keys that have cached metadata. */
  getCachedPlaylists(): string[] {
    return Object.keys(this.data.playlists);
  }

  /** Get the cached ETag for a playlist, or null if not cached. */
  getETag(playlistKey: string): string | null {
    return this.data.playlists[playlistKey]?.etag ?? null;
  }

  // ── Write ──

  /** Store a playlist file list with its ETag. */
  storePlaylistFiles(
    playlistKey: string,
    files: FileInfo[],
    etag: string | null,
    name?: string,
  ): void {
    this.data.playlists[playlistKey] = {
      files,
      etag,
      playlistName: name,
      fileCount: files.length,
      cachedAt: new Date().toISOString(),
    };
    this.persist();
  }

  /** Remove cached data for a playlist. */
  removePlaylist(playlistKey: string): void {
    if (this.data.playlists[playlistKey]) {
      delete this.data.playlists[playlistKey];
      this.persist();
    }
  }

  /** Clear all cached metadata. */
  clearAll(): void {
    this.data.playlists = {};
    this.persist();
  }

  // ── Internal ──

  private persist(): void {
    saveJsonIndex(this.cachePath, this.data);
  }
}

File: ios/MusicPorter/MusicPorter/Services/Cache/MetadataCache.swift
import Foundation

/// Persistent cache for API response metadata (playlist file lists + ETags).
///
/// Stores one JSON file per profile at <Application Support>/MusicPorter/cache/<profile>/metadata-cache.json.
/// Separate from the audio cache index (cache-index.json).
///
/// Port of sync client's metadata-cache.ts. Both implementations must use identical
/// JSON formats, schema versions, and cache invalidation behavior.
actor MetadataCache {
    private let cachePath: URL
    private var data: MetadataCacheData

    init(profile: String) {
        let cacheDir = CacheUtils.cacheDirectory(profile: profile)
        let fm = FileManager.default
        if !fm.fileExists(atPath: cacheDir.path) {
            try? fm.createDirectory(at: cacheDir, withIntermediateDirectories: true)
        }
        self.cachePath = cacheDir.appendingPathComponent(CacheConstants.metadataCacheFilename)
        self.data = CacheUtils.loadJsonIndex(
            path: cachePath,
            fallback: MetadataCacheData(
                profile: profile,
                version: CacheConstants.metadataCacheVersion,
                playlists: [:]
            ),
            validator: { $0.profile == profile && $0.version == CacheConstants.metadataCacheVersion }
        )
    }

    // MARK: - Read

    /// Get cached playlist data, or nil if not cached.
    func getPlaylistFiles(_ playlistKey: String) -> CachedPlaylistData? {
        data.playlists[playlistKey]
    }

    /// Get list of playlist keys that have cached metadata.
    func getCachedPlaylists() -> [String] {
        Array(data.playlists.keys)
    }

    /// Get the cached ETag for a playlist, or nil if not cached.
    func getETag(_ playlistKey: String) -> String? {
        data.playlists[playlistKey]?.etag
    }

    // MARK: - Write

    /// Store a playlist file list with its ETag.
    func storePlaylistFiles(
        _ playlistKey: String,
        files: [CachedFileInfo],
        etag: String?,
        name: String? = nil
    ) {
        data.playlists[playlistKey] = CachedPlaylistData(
            files: files,
            etag: etag,
            playlistName: name,
            fileCount: files.count,
            cachedAt: CacheUtils.isoNow()
        )
        persist()
    }

    /// Remove cached data for a playlist.
    func removePlaylist(_ playlistKey: String) {
        guard data.playlists[playlistKey] != nil else { return }
        data.playlists.removeValue(forKey: playlistKey)
        persist()
    }

    /// Clear all cached metadata.
    func clearAll() {
        data.playlists = [:]
        persist()
    }

    // MARK: - Internal

    private func persist() {
        CacheUtils.saveJsonIndex(path: cachePath, data: data)
    }
}
